agent/cache-types/streaming_health_services.go

```diff
@@ -0,0 +1,187 @@
+package cachetype
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-bexpr"
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/consul/agent/cache"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/agent/submatview"
+	"github.com/hashicorp/consul/lib/retry"
+	"github.com/hashicorp/consul/proto/pbservice"
+	"github.com/hashicorp/consul/proto/pbsubscribe"
+)
+
+const (
+	// Recommended name for registration.
+	StreamingHealthServicesName = "streaming-health-services"
+)
+
+// StreamingHealthServices supports fetching discovering service instances via the
+// catalog using the streaming gRPC endpoint.
+type StreamingHealthServices struct {
+	RegisterOptionsBlockingRefresh
+	deps MaterializerDeps
+}
+
+// NewStreamingHealthServices creates a cache-type for watching for service
+// health results via streaming updates.
+func NewStreamingHealthServices(deps MaterializerDeps) *StreamingHealthServices {
+	return &StreamingHealthServices{deps: deps}
+}
+
+type MaterializerDeps struct {
+	Client submatview.StreamClient
+	Logger hclog.Logger
+}
+
+// Fetch implements cache.Type
+func (c *StreamingHealthServices) Fetch(opts cache.FetchOptions, req cache.Request) (cache.FetchResult, error) {
+	if opts.LastResult != nil && opts.LastResult.State != nil {
+		return opts.LastResult.State.(*streamingHealthState).Fetch(opts)
+	}
+
+	srvReq := req.(*structs.ServiceSpecificRequest)
+	newReqFn := func(index uint64) pbsubscribe.SubscribeRequest {
+		req := pbsubscribe.SubscribeRequest{
```

Comment:
That makes sense. Namespaces are not hooked up at all yet, and I have an item on the project plan to add it next. There are TODOs in the `agent/consul/stream` package for receiving the namespace as well.

What I'd like to do is add a `Namespace` to `SubscribeRequest`, so that we don't need different versions for oss and ent.

I'll add a `TODO(streaming)` where so we don't forgot this spot.